awslambdaric/lambda_runtime_log_utils.py

```diff
@@ -0,0 +1,98 @@
+"""
+Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+"""
+
+import logging
+
+from .lambda_runtime_marshaller import to_json
+
+RESERVED_FIELDS = {
+    "name",
+    "msg",
+    "args",
+    "levelname",
+    "levelno",
+    "pathname",
+    "filename",
+    "module",
+    "exc_info",
+    "exc_text",
+    "stack_info",
+    "lineno",
+    "funcName",
+    "created",
+    "msecs",
+    "relativeCreated",
+    "thread",
+    "threadName",
+    "processName",
+    "process",
+    "aws_request_id",
+}
+
+
+class JsonFormatter(logging.Formatter):
+    def __init__(self):
+        super().__init__(datefmt="%Y-%m-%dT%H:%M:%SZ", validate=False)
+
+    def format_stacktrace(self, exc_info):
+        if not exc_info:
+            return None
+        return self.formatException(exc_info)
+
+    @staticmethod
+    def format_exception_name(exc_info):
+        if not exc_info:
+            return None
+
+        return exc_info[0].__name__
+
+    @staticmethod
+    def format_exception(exc_info):
+        if not exc_info:
+            return None
+
+        return str(exc_info[1])
+
+    @staticmethod
+    def format_location(record: logging.LogRecord):
+        if not record.exc_info:
+            return None
+
+        return f"{record.pathname}:{record.funcName}:{record.lineno}"
+
+    def format(self, record: logging.LogRecord) -> str:
+        result = {
+            "timestamp": self.formatTime(record, self.datefmt),
+            "level": record.levelname,
+            "message": record.getMessage(),
+            "logger": record.name,
+            "stackTrace": self.format_stacktrace(record.exc_info),
+            "errorType": self.format_exception_name(record.exc_info),
+            "errorMessage": self.format_exception(record.exc_info),
+            "requestId": getattr(record, "aws_request_id", None),
+            "location": self.format_location(record),
+        }
+        result.update(
+            (key, value)
+            for key, value in record.__dict__.items()
+            if key not in RESERVED_FIELDS and key not in result
+        )
+        result.update(INSTANCE_LOGGING_CONTEXT)
+        result.update(INVOCATION_LOGGING_CONTEXT)
+
+        result = {k: v for k, v in result.items() if v is not None}
+
+        return to_json(result) + "\n"
+
+
+INSTANCE_LOGGING_CONTEXT = dict()
+INVOCATION_LOGGING_CONTEXT = dict()
+
+
+def lambda_instance_logs_update(key: str, value: str):
+    INSTANCE_LOGGING_CONTEXT[str(key)] = str(value)
```

Comment:
why `str(value)`? I think the value should not be stringified - customer may want to attach an object/dict there that will be rendered as such in JSON